symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Optimization.hs

```diff
@@ -0,0 +1,99 @@
+module ZkFold.Symbolic.Compiler.ArithmeticCircuit.Optimization (
+        optimize,
+        toConstVar
+    ) where
+
+import           Data.Bool                                               (bool)
+import           Data.Functor                                            ((<&>))
+import           Data.Functor.Rep                                        (Representable (..))
+import           Data.Map                                                hiding (drop, foldl, foldr, map, null, splitAt,
+                                                                          take)
+import qualified Data.Map.Internal                                       as M
+import qualified Data.Map.Monoidal                                       as MM
+import           Data.Maybe                                              (catMaybes)
+import qualified Data.Set                                                as S
+import           Numeric.Natural                                         (Natural)
+import           Prelude                                                 hiding (Num (..), drop, length, product,
+                                                                          splitAt, sum, take, (!!), (^))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..))
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..), polynomial)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+
+--------------------------------- High-level functions --------------------------------
+
+-- | Replaces linear polynoms of the form
+-- @(fromConstant k) * (NewVar nV) + (fromConstant c)@
+-- with a constant variable @fromConstant $ negate c // k@ in an arithmetic circuit
+-- and replaces variable with a constant in witness
+--
+optimize ::
+  (Arithmetic a, Ord (Rep i), Functor o) =>
+  ArithmeticCircuit a p i o  -> ArithmeticCircuit a p i o
+optimize ac = let (ac', vs) = varsToReplace (ac, []) in foldr optimizeAc ac' vs
+
+varsToReplace ::
+  (Arithmetic a, Ord (Rep i), Functor o) =>
+  (ArithmeticCircuit a p i o , [(SysVar i, a)]) -> (ArithmeticCircuit a p i o , [(SysVar i, a)])
+varsToReplace (ac, l) = case catMaybes [toConstVar p | (_, p) <- toList (acSystem ac)] of
+    [] -> (ac, l)
+    vs -> varsToReplace (ac {acSystem = acSystem $ foldr optimizeSystems ac vs}, l ++ vs)
+
+optimizeSystems :: forall a p i o. (Arithmetic a, Ord (Rep i)) =>
```

Comment:
Since you're only modifying the system, it would be cleaner to only use a system as input/output instead of the whole circuit